FC-73 Implementing xqueue_submission service to send student response to edx-submission

gabrielC1409 opened the pull request:
## Description
This pull request modifies the student submission process to route submissions through the edx-submission service, replacing the traditional XQueue service.

## Key Changes:

* Student Submission Routing: Submissions are now directed to the edx-submission service, providing a more modern and efficient architecture for handling student responses.

* Waffle Flag Implementation: A new waffle flag, send_to_submission_course.enable, has been introduced in the Django admin interface. This flag allows administrators to enable or disable the edx-submission functionality for specific courses without modifying the codebase.


## Modifications in xqueue_interfaces.py:

Service Selection: The system now checks the state of the waffle flag for each course. If enabled, student responses are directed to the edx-submission service via the send_to_submission function; otherwise, they continue to be sent to XQueue.​
Callback URL Construction: The construct_callback method in the XQueueService class has been updated to generate appropriate callback URLs based on the waffle flag's state, ensuring that evaluation results are returned correctly.​
Addition of send_to_submission in xqueue_submission.py: This function processes student responses by extracting relevant data (e.g., course_id, item_id, student_id, max_score) and interacts with the edx-submission API to create new submission records, ensuring proper storage and processing.​

## Motivation
The primary goal is to leverage the edx-submission service's modern architecture for handling student submissions, offering improved efficiency and scalability. The introduction of a course-specific waffle flag facilitates a controlled and flexible transition, allowing for A/B testing and gradual adoption.​

## Testing
Unit tests in test_xqueue_interface.py and test_xqueue_submission.py have been updated to cover the new functionality. Additionally, manual testing was conducted to verify that submissions are correctly routed based on the waffle flag's state and that the system behaves as expected under various configurations.​

## Additional Considerations

⚠️Warning: The XQueue service is planned for deprecation in favor of the edx-submission service. This waffle flag serves as a temporary measure to facilitate the transition. Future updates will remove the flag, making edx-submission the default submission handler.

## Recommendations for Review
Code Review: Examine the modifications in xqueue_interfaces.py and xqueue_submission.py to ensure the logic aligns with the intended functionality and adheres to coding standards.​

Testing: Review and, if necessary, expand the test cases in test_xqueue_interface.py and test_xqueue_submission.py to ensure comprehensive coverage of the new feature.​

Documentation: Confirm that the implementation details and usage instructions for the waffle flag are clearly documented, facilitating ease of use for administrators.

openedx-webhooks commented:
Thanks for the pull request, @gabrielC1409!


This repository is currently maintained by `@openedx/wg-maintenance-edx-platform`.


Once you've gone through the following steps feel free to tag them in a comment and let them know that your changes are ready for engineering review.

<details><summary>:radio_button: Get product approval</summary>

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.
</details>

<details><summary>:radio_button: Provide context</summary>

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads
</details>





<details><summary>:radio_button: Get a green build</summary>

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.
</details><details>


---

<details><summary>Where can I find more information?</summary>

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)
</details>

<details><summary>When can I expect my changes to be merged?</summary>

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*
</details>
<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

sarina commented on docs/decisions/0022-create-waffle-switch.rst:
```diff
@@ -0,0 +1,58 @@
+###############################################################
+Integration of Waffle Switch for XQueue Submission
```
I think this ADR should be in the https://github.com/openedx/edx-platform/tree/58834c6bdc20fae182eca4aa4ce7402cc14dfd47/xmodule/docs/decisions directory, and the file name should include "xqueue" in it (we have a ton of waffle switches, so the title "create-waffle-switch" is really not specific enough)

sarina commented:
Your conventional commit names are inaccurate. A new ADR should use the `docs:` prefix and a new feature `feat:`. Please read through https://docs.openedx.org/projects/openedx-proposals/en/latest/best-practices/oep-0051-bp-conventional-commits.html

sarina commented:
@gabrielC1409 every time you push a commit I need to come into the PR and manually re-run the tests. You won't get a different result on the test if you just keep merging in master. I will next run tests for you when you push a new commit of code.

sarina commented:
@gabrielC1409 this comment still applies, can you acknowledge that you've seen it?

> Your conventional commit names are inaccurate. A new ADR should use the docs: prefix and a new feature feat:. Please read through https://docs.openedx.org/projects/openedx-proposals/en/latest/best-practices/oep-0051-bp-conventional-commits.html

gabrielC1409 commented:
hi @sarina how are you? I already have the updated changes, I set the PR as open, I am waiting for your feedback.

sarina commented:
@gabrielC1409 you need to reword your commit messages. They are not using the correct Conventional Commit prefixes. cc @angonz

sarina commented:
Also on your last commit you had a failing test. All tests need to pass, and the conventional commit prefixes need to be accurate, before we review.

angonz commented:
@gabrielC1409 please reset the PR to draft until we can review internally.

angonz commented:
@gabrielC1409 I have cleaned up some changes and squashed commits. Please check the use of dateformat and how are you checking the course waffle flag and resubmit.

angonz commented on xmodule/capa/xqueue_interface.py:
```diff
@@ -26,6 +28,34 @@
+    return flag and flag.enabled
+
+
+def get_flag_by_name(flag_name):
```
I think you don't need this function any more

angonz commented on xmodule/capa/xqueue_interface.py:
```diff
@@ -26,6 +28,34 @@
+    return flag and flag.enabled
+
+
+def get_course_id(callback_url):
```
I think that with the new approach you don't need this function

angonz commented on xmodule/capa/xqueue_interface.py:
```diff
@@ -76,11 +106,15 @@ class XQueueInterface:
-    def __init__(self, url: str, django_auth: Dict[str, str], requests_auth: Optional[HTTPBasicAuth] = None):
+    def __init__(self, url: str, django_auth: Dict[str, str],
+                 requests_auth: Optional[HTTPBasicAuth] = None,
+                 block: 'ProblemBlock' = None):
```
Please add a comment to document that we are adding `block` as a parameter only to extract the course_id to check the course waffle flag. This can be removed after the legacy xqueue is deprecated.

gabrielC1409 replied:
Okay @angonz, I've already uploaded the suggested changes.

angonz commented on xmodule/docs/decisions/0005-send-data-to-edx-submission.rst:
```diff
@@ -0,0 +1,194 @@
+#########################################################
```
Let's review this decision together before submitting.

angonz reviewed (changes requested):
Please check comments and let's review the decision doc together

angonz commented on xmodule/docs/decisions/0005-send-data-to-edx-submission.rst:
```diff
@@ -0,0 +1,194 @@
+Context
+*******
+
+On the Open edX platform, student responses to assignments are traditionally submitted to the `XQueue` service for assessment. However, a need was identified to allow certain responses to be submitted to the `edx-submission` service, which offers a more modern and efficient architecture for handling submissions. To facilitate a controlled transition and allow for A/B testing, the introduction of  ...
```
Instead of 

> On the Open edX platform, student responses to assignments are traditionally submitted to the `XQueue` service for assessment.

I would say "`coderesponse` problems configured for external grading are traditionally submitted to an XQueue server for assessment."

angonz replied:
Focus on the main objective of the PR, which is to send submissions to edx-submissions so that XQueue server can be deprecated.

angonz commented on xmodule/docs/decisions/0005-send-data-to-edx-submission.rst:
```diff
@@ -0,0 +1,194 @@
+
+   The updated logic in `construct_callback` involves checking the state of the waffle flag and constructing the callback URL accordingly. If the flag is active for a given course, the callback URL is constructed for `edx-submission`; otherwise, it is constructed for `XQueue`.
+
+4. **Implementation of `send_to_submission` in `xqueue_submission.py`**: The `send_to_submission` function was developed in the `xqueue_submission.py` file. This function is responsible for:
```
This is # 1

e0d commented:
@gabrielC1409 I've invited you the the triage team so your tests will run automatically.  Please accept that invitation.

gabrielC1409 commented:
Hi @e0d how are you? Where did you send me that invitation? I can't see it by email to accept the invitation.

> @gabrielC1409 I've invited you the the triage team so your tests will run automatically. Please accept that invitation.

gabrielC1409 commented:
> Hi @e0d how are you? Where did you send me that invitation? I can't see it by email to accept the invitation.
> 
> > @gabrielC1409 I've invited you the the triage team so your tests will run automatically. Please accept that invitation.

@e0d  This is my email so I can receive the invitation:

gcarvajal@aulasneo.com

e0d commented:
The invitation will come to whatever email is associated with your github account.

sarina commented:
>  They can also visit https://github.com/openedx to accept the invitation.

@gabrielC1409 please visit the above link.

gabrielC1409 commented:
> The invitation will come to whatever email is associated with your github account.

Okay, I already accepted the invitation.

![Captura de pantalla 2025-03-06 102535](https://github.com/user-attachments/assets/b2859155-9163-43b1-997e-b711b4fb3945)

angonz reviewed (changes requested):
Fix double "Accepted" in decision doc, squash commits and push again

gabrielC1409 commented:
Hi @sarina , how are you? I'm leaving this comment to let you know that the PR should be set as open. We've already done the code review internally and the tests passed successfully.

sarina commented on scripts/xsslint_thresholds.json:
```diff
@@ -8,7 +8,7 @@
         "django-trans-escape-variable-mismatch": 0,
         "django-trans-invalid-escape-filter": 0,
-        "django-trans-missing-escape": 0,
+        "django-trans-missing-escape": 0, 
```
Revert changes to this file, pleas.

sarina commented on xmodule/capa/errors.py:
```diff
@@ -0,0 +1,57 @@
+
+class XQueueSubmissionError(Exception):
+    """Base class for all XQueue submission errors."""
+    # No es necesario el `pass`, la clase ya hereda de Exception.
```
Please update this to be in English

sarina commented on xmodule/capa/errors.py:
```diff
@@ -0,0 +1,57 @@
+
+class GetSubmissionParamsError(XQueueSubmissionError):
+    """Raised when there is an issue getting submission parameters."""
+    MESSAGE = "Block instance is not defined!"
```
This `MESSAGE` doesn't match the docstring, in my mind. Could you explain a bit more how this message relates to the docstring?

gabrielC1409 replied:
This message is related to the use when a response is sent and the generated response is not of type xblock

sarina commented on xmodule/capa/inputtypes.py:
```diff
@@ -408,7 +406,7 @@ class OptionInput(InputTypeBase):
     Example:
 
-    <optioninput options="('Up','Down')" correct="Up"/><text>The location of the sky</text>
+    <optioninput options="('Up', 'Down')" correct="Up"/><text>The location of the sky</text>
```
Why whitespace changes, here and in the rest of the file? (Changing whitespace is usually unnecessary, and complicates `git blame` in the future) I don't think any of the changes in this file are needed.

gabrielC1409 replied:
I had to add that space because when running the tests directly in the xmodule/capa/test folder some files asked me to make that correction.

sarina commented on xmodule/capa/xqueue_interface.py:
```diff
@@ -73,14 +83,26 @@ def parse_xreply(xreply):
+        requests_auth (Optional[HTTPBasicAuth], optional): Authentication for HTTP requests. Defaults to None.
+        block ('ProblemBlock', optional): Added as a parameter only to extract the course_id
+        to check the course waffle flag `send_to_submission_course.enable`.
+        This can be removed after the legacy xqueue is deprecated. Defaults to None.
```
I think this is a docstring for the `__init__` function, not the class itself

sarina commented on xmodule/capa/xqueue_submission.py:
```diff
@@ -0,0 +1,112 @@
+            log.error("%s", e)
+            return {"error": str(e)}
+
+        except ValidationError as e:
```
These three exceptions do the same thing. change to catch all 3 exceptions in the same clause. See https://docs.python.org/3/tutorial/errors.html#handling-exceptions

sarina commented on xmodule/docs/decisions/0005-send-data-to-edx-submission.rst:
```diff
@@ -0,0 +1,191 @@
+Context
+*******
+
+Traditionally, the Open edX platform has utilized the `XQueue` service to handle student assignment submissions. However, to leverage a more modern and efficient architecture, there is a need to transition these submissions to the `edx-submission` service. This transition aims to improve the scalability and maintainability of the submission handling process.
```
Wrap code in double backticks please, throughout the RST file

```suggestion
Traditionally, the Open edX platform has utilized the ``XQueue`` service to handle student assignment submissions. However, to leverage a more modern and efficient architecture, there is a need to transition these submissions to the ``edx-submission`` service. This transition aims to improve the scalability and maintainability of the submission handling process.
```

sarina commented on xmodule/docs/decisions/0005-send-data-to-edx-submission.rst:
```diff
@@ -0,0 +1,191 @@
+The following changes have been implemented to redirect student submissions to the `edx-submission` service:
+
+1. **Introduction of `send_to_submission` Function in `xqueue_submission.py`:**
+   - **Functionality:** A new function, `send_to_submission`, has been developed to process student responses. It parses essential information such as `course_id`, `item_id`, `student_id`, and `max_score` from the submission data.
```
Please add blank lines above and below each bullet. Check the rendered file - you can see currently this renders as a giant block instead of a bulleted list. https://github.com/openedx/edx-platform/blob/ef018510eca65c47d26a4a60975b518820f04634/xmodule/docs/decisions/0005-send-data-to-edx-submission.rst

sarina commented on xmodule/docs/decisions/0005-send-data-to-edx-submission.rst:
```diff
@@ -0,0 +1,191 @@
+   - **Deprecation Plan:** This *waffle flag* is a temporary measure to facilitate a smooth transition. Once the `edx-submission` service is fully adopted and validated, the flag will be deprecated, and all submissions will be processed exclusively through `edx-submission`.
+
+
+Configuration for Xqueue-watcher:
```
@ormsbee - do you think this would be more appropriate somewhere in a `README` file rather than within an ADR?

ormsbee replied:
I think that it makes sense to eventually put it in the xqueue-watcher README once the entirety of this merged into master, but it's fine to stay in this ADR for now.

sarina commented on xmodule/docs/decisions/0005-send-data-to-edx-submission.rst:
```diff
@@ -0,0 +1,191 @@
+   - **Deprecation Plan:** This *waffle flag* is a temporary measure to facilitate a smooth transition. Once the `edx-submission` service is fully adopted and validated, the flag will be deprecated, and all submissions will be processed exclusively through `edx-submission`.
+
+
+Configuration for Xqueue-watcher:
```
```suggestion
Configuration for Xqueue-watcher:
***************************************
```

sarina commented on xmodule/docs/decisions/0005-send-data-to-edx-submission.rst:
```diff
@@ -0,0 +1,191 @@
+References
+**********
+
+- **Relevant commits**: [Implementation of the Waffle Flag and modification of xqueue_interfaces.py](https://github.com/aulasneo/edx-platform/commit/f50afcc301bdc3eeb42a 6dc2c051ffb2d799f868#diff-9b4290d2b574f54e4eca7831368727f7ddbac8292aa75ba4b28651d4bf2bbe6b)
```
This is RST, not markdown. See https://sublime-and-sphinx-guide.readthedocs.io/en/latest/references.html#links-to-external-web-pages

Also I am not sure it's appropriate to link to a single commit on a fork. Linking to this PR may be more appropriate.

sarina reviewed (commented):
@ormsbee please put this on your review queue

gabrielC1409 committed:
feat: Send xqueue submissions to edx-submission

fix: Restructuring to send course_id and score to edx submission

fix: Refactoring of sending information to sent_to_submission

fix: Elimination of unnecessary functions

fix: Added usage comment to ProblemBlock in XqueueInterface constructor

fix: update doc ADR

fix: setting for Quality Others test (ubuntu-24.04, 3.11, 20)

fix: Deprecation for django-trans-escape-filter-parse-error

test: Add @pytest.mark.django_db decorator to test functions

test: Fix for pylint being disabled

fix: updated changes for pylint disable

fix: update error from docs ADR-0005

update: xmodule/docs/decisions/0005-send-data-to-edx-submission.rst

Co-authored-by: Sarina Canelake <sarina@axim.org>

update: xmodule/docs/decisions/0005-send-data-to-edx-submission.rst

Co-authored-by: Sarina Canelake <sarina@axim.org>

fix: Adjusted correction

fix: update date for docs ADR

Revert "fix: update date for docs ADR"

This reverts commit 0b4229c51c4937f95cb407872645dd448df45418.

fix: replace call created_submission to create_external_grader_detail

fix: update test xqueue_submission

fix: add docstring in test_xqueue_submission

fix: update date doc ADR

fix: update version edx-submission 3.8.6

fix: add @pytest.mark.xfail

fix: add 20 chances in test_capa_block:

fix: increase retry attempts for seed generation in ProblemBlockTest

fix: change version to edx-submission lib

fix: new version edx-submission in testings

fix: replace parameter file to files

fix: update variable grader_file_name and points_possible

fix: Adjustment in the is_flag_active function to always take the last record edited in the waffle

fix: wrap large line of code

fix: update function is_flag_active

fix: code style adjustment

fix: changes for 60 retry

feat: use CourseWaffleFlag to determine xqueue callback path

fix: Code style adjustment

fix: remove deprecated xqueue callback route and simplify callback type logic

fix: Deleting a comment in the ADR document

fix: add log in self.block is None

fix: Code style adjustment in log

gabrielC1409 commented:
Hi @angonz , how are you? I've already uploaded the changes to the PR. When you can, I'll do a code review to validate it.

angonz commented on xmodule/capa/xqueue_submission.py:
```diff
@@ -0,0 +1,102 @@
+
+        try:
+            student_item, answer, queue_name, grader, score = self.get_submission_params(header, body)
+            return create_external_grader_detail(  # pylint: disable=unexpected-keyword-arg
```
Remove the pylint disable

angonz commented on requirements/edx/base.txt:
```diff
@@ -524,7 +524,7 @@ edx-search==4.1.3
 edx-sga==0.25.0
     # via -r requirements/edx/bundled.in
-edx-submissions==3.8.5
+edx-submissions==3.8.6
```
Please double check these dependencies

sarina replied:
For testing purposes, you should be pinning edx-submissions to the branch for the other PRs. Otherwise you're not testing the new changes.

angonz reviewed (approved):
Ok on my side. Go ahead open the PR for review.

gabrielC1409 commented:
Hi @ormsbee , how are you?

I wanted to comment that I added the following dependency in the requirements files:

`edx-submissions @ git+https://github.com/aulasneo/edx-submissions.git@XQU-15-add-fields-to-submission-model`

This was necessary because I'm working on a branch (XQU-15-add-fields-to-submission-model) that hasn't yet been merged into master. The changes needed for this PR are on that branch, so I need to point it directly to the Git repository so the functionality is available during development and testing.

Once the other PR is approved and merged, this dependency can be replaced with the corresponding version from PyPI.

I'm here for any comments or adjustments you deem necessary.

![image](https://github.com/user-attachments/assets/9efae9cb-34ea-4120-93fb-ff2c872a5d52)

ormsbee commented on xmodule/capa/xqueue_interface.py:
```diff
@@ -26,6 +27,27 @@
+    if latest and latest.enabled and latest.override_choice:
+        return True
+
+    return False
```
Please use `CourseWaffleFlag` instead of implementing this yourself.

https://github.com/openedx/edx-platform/blob/fb62eaf94ce5b391137956550b184c9dc89e5a42/openedx/core/djangoapps/waffle_utils/__init__.py#L13

gabrielC1409 replied:
Ok, I'll let you know when this setting is implemented.

ormsbee replied:
Note that `CourseWaffleFlag.is_enabled` takes a `CourseKey` and not a string.

ormsbee commented on xmodule/capa/xqueue_interface.py:
```diff
@@ -135,7 +170,17 @@ def _send_to_queue(self, header, body, files_to_upload):  # lint-amnesty, pylint
                 files.update({f.name: f})
 
-        return self._http_post(self.url + '/xqueue/submit/', payload, files=files)
+        course_id = str(self.block.scope_ids.usage_id.context_key)
```
If `self.block` can technically be `None`, please do a check for that (i.e. log an error in that case and fall back to the legacy behavior).

ormsbee commented on xmodule/capa/xqueue_interface.py:
```diff
@@ -135,7 +170,17 @@ def _send_to_queue(self, header, body, files_to_upload):  # lint-amnesty, pylint
+        if is_flag_active('send_to_submission_course.enable', course_id):
+            # Use the new edx-submissions workflow
+            submission = self.submission.send_to_submission(header, body, files)
+            log.error(submission)
```
Is this leftover from debugging?

ormsbee commented on xmodule/capa/xqueue_submission.py:
```diff
@@ -0,0 +1,104 @@
+        """
+        Submits the extracted student data to the edx-submissions system.
+        """
+        from submissions.api import create_external_grader_detail
```
Why is this a local import, as opposed to being at the top of the module?

ormsbee commented on xmodule/capa/xqueue_submission.py:
```diff
@@ -0,0 +1,104 @@
+    def _parse_json(self, data, name):
+        """Helper function to parse JSON safely."""
+        try:
+            return json.loads(data) if isinstance(data, str) else data
```
Please explain in a docstring why this is necessary, i.e. why do we not know whether we should be parsing something as JSON or sending it back as-is?

ormsbee commented on xmodule/capa/xqueue_submission.py:
```diff
@@ -0,0 +1,104 @@
+        block_id = self.block.scope_ids.def_id.block_id
+        points_possible = self.block.max_score()
+
+        item_id = f"block-v1:{course_id.replace('course-v1:', '')}+type@{item_type}+block@{block_id}"
```
Never build block usage keys by hand, because it will break in various edge cases that will be hard for you to test, e.g. CCX courses that use a different prefix.

In this case though, why do we need to do this? it looks like we're just reconstructing `self.block.usage_key`?

ormsbee replied:
If you do need to create a UsageKey manually, you can do so by grabbing a LearningContextKey (in this case a CourseKey) and calling `.make_usage_key()` on it with the appropriate parameters.

ormsbee reviewed (changes requested):
There are capa/modulestore internals that I'm trying to dig through, so some of this feedback is going to be slower than I had hoped. I'm probably going to do 1-2 more of these smaller-chunked reviews on pieces of this.

ormsbee commented on xmodule/capa/xqueue_interface.py:
```diff
@@ -25,6 +27,32 @@
+# .. toggle_will_remain_in_codebase: True
+# .. toggle_tickets: none
+# .. toggle_status: supported
+SEND_TO_SUBMISSION_COURSE_FLAG = CourseWaffleFlag('send_to_submission_course.enable', __name__)
```
Please rename this to `USE_EDX_SUBMISSIONS_FOR_XQUEUE` / `use_edx_submissions_for_xqueue`.

ormsbee commented on xmodule/capa/xqueue_interface.py:
```diff
@@ -25,6 +27,32 @@
+# .. toggle_default: False
+# .. toggle_use_cases: opt_in
+# .. toggle_creation_date: 2024-04-03
+# .. toggle_expiration_date: None
```
Expiration date should be 2025-08-12, the date currently listed in the [DEPR ticket](https://github.com/openedx/public-engineering/issues/286).

ormsbee commented on xmodule/capa/xqueue_interface.py:
```diff
@@ -25,6 +27,32 @@
+SEND_TO_SUBMISSION_COURSE_FLAG = CourseWaffleFlag('send_to_submission_course.enable', __name__)
+
+
+def is_flag_active(flag_name, course_id):
```
The purpose of having this type of helper function is to insulate the rest of the code from changes in how we want to roll out a feature. Sometimes a feature goes from being opt-in to being opt-out, sometimes we gate the feature entirely with an operator-level Django setting, and sometimes those requirements change over time. You can see some examples of that here:

https://github.com/openedx/edx-platform/blob/master/lms/djangoapps/courseware/toggles.py

Also, most of the time, we should have an actual CourseKey in hand when calling this function, since even if we're taking user input, we'd probably parse it to make sure it's a valid one.

Please do the following:

1. Rename this to something that describes the decision being made, e.g. `def use_edx_submissions_for_xqueue` or something along those lines, rather than a question about whether a specific toggle is on or off.
2. Don't take a flag_name as an argument. The specific flags that this function looks at are an implementation detail and may change over time. The application code should be insulated from that.
3. Please have it take a course_key instead of a course_id, and use type annotations to show it's a CourseKey object.
4. Make it so that the course_key argument is optional and defaults to `None`. That way, app code can check to see if it's enabled at the site level more clearly without having to send a CourseKey at all. (CourseWaffleFlag.is_enabled() supports sending in None for  CourseKey, so you shouldn't have to do any special handling).

ormsbee commented on xmodule/capa/errors.py:
```diff
@@ -0,0 +1,56 @@
+    MESSAGE = "Validation error: {error}"
+
+    def __init__(self, error):
+        super().__init__(self.MESSAGE.format(error=error))
```
Question (not requesting a change at the moment): What is the value of having granular errors that subclass XQueueSubmissionError if we never catch XQueueSubmissionError? Why use some of these instead of built-in errors like `KeyError` or `ValueError`? Is it mostly to make the test code clearer?

gabrielC1409 replied:
The main goal of these custom errors is to separate exceptions from the XQueue domain from the rest of the system. Although we aren't currently directly capturing XQueueSubmissionError in many places, it allows us to:

Have a clear XQueue-specific error hierarchy, which facilitates logging, debugging, and future error handling workflows.

Allow other services (or tests) to capture XQueue errors generically, without worrying about finer details like KeyError or JSONDecodeError.

Make error messages more descriptive and friendly to logging and debugging.

ormsbee replied:
Okay. I personally think this is a bit of overkill at the moment, but I'm okay with leaving it in. Thank you for the explanation.

ormsbee commented on xmodule/capa/xqueue_interface.py:
```diff
@@ -180,21 +232,32 @@ def interface(self):
+        userid = str(self._block.scope_ids.user_id)
+        mod_id = str(self._block.scope_ids.usage_id)
+
+        callback_type = (
```
I'm a little confused... what calls the callback URL later on when it's going through edx-submissions? In the legacy experience, the xqueue service calls this URL after xqueue-watcher submits the answer to xqueue, right? But if xqueue-watcher is submitting to edx-submissions, what would call the callback URL constructed in this function?

gabrielC1409 replied:
Here's how the callback logic works:

The method `construct_callback()` dynamically builds the callback URL based on whether the course is using the new `edx-submissions` flow or the legacy `xqueue` flow.

- If `use_edx_submissions_for_xqueue(course_key)` returns **True**, then the constructed URL will point to the **`callback_submission`** endpoint. This is used when the course is configured to use the new `edx-submissions` pipeline for grading.

- If the flag returns **False**, the URL will instead use the **`xqueue_callback`** endpoint, which is what the legacy xqueue-watcher calls after grading is complete.

So the purpose of this logic is to decouple the callback construction from the submission implementation — ensuring that the correct LMS endpoint is called depending on the toggle state (`send_to_submission_course.enable`).

ormsbee replied:
Sorry, I might need you to just walk me through this step by step....

The callback_submission endpoint is still this, right?

```python
    re_path(
        r'^courses/{}/xqueue/(?P<userid>[^/]*)/(?P<mod_id>.*?)/(?P<dispatch>[^/]*)$'.format(
            settings.COURSE_ID_PATTERN,
        ),
        xqueue_callback,
        name='callback_submission',
    ),
```

But if we're using edx-submissions, nothing would ever POST to that URL, would it? In the legacy flow, the xqueue service would have posted to that URL in the LMS after xqueue-watcher POSTed to xqueue. In the new flow, xqueue-watcher POSTs to the edx-submissions provided endpoint, which is already in the LMS, and we update the score by listening to the signal that gets sent by edx-submissions. So why would anything POST to this URL in the new flow?

gabrielC1409 replied:
You're right: in the edx-submissions flow, there's no longer anything that POSTs callback_submission.

The URL is still constructed for compatibility reasons or because a common pattern is being used for both flows, but in practice, this endpoint isn't used in the new flow.

ormsbee commented on xmodule/capa/xqueue_interface.py:
```diff
@@ -135,6 +179,16 @@ def _send_to_queue(self, header, body, files_to_upload):  # lint-amnesty, pylint
 
+        if self.block is None:
+            # XQueueInterface: if self.block is None, falling back to legacy xqueue submission.
+            return self._http_post(self.url + '/xqueue/submit/', payload, files=files)
```
I think falling back to the old behavior in this case is good. But since we aren't expecting to ever get into this state, please log an error here so that people don't think their transition is working when it's not.

ormsbee replied:
@gabrielC1409: Sorry, I forgot to check this before giving a ✅ -- can you please add this logging before I merge?

ormsbee replied:
@gabrielC1409: Sorry, I forgot to check this before giving a ✅ -- can you please add this logging before I merge?

ormsbee commented on xmodule/capa/xqueue_interface.py:
```diff
@@ -180,21 +235,28 @@ def interface(self):
         """
         return self._interface
 
-    def construct_callback(self, dispatch: str = 'score_update') -> str:
+    def construct_callback(self, dispatch: str = "score_update") -> str:
         """
-        Return a fully qualified callback URL for external queueing system.
+        Return a fully qualified callback URL for the external queueing system.
         """
+        course_key = self._block.scope_ids.usage_id.context_key
+        userid = str(self._block.scope_ids.user_id)
+        mod_id = str(self._block.scope_ids.usage_id)
+
+        callback_type = "xqueue_callback"
+
         relative_xqueue_callback_url = reverse(
-            'xqueue_callback',
-            kwargs=dict(
-                course_id=str(self._block.scope_ids.usage_id.context_key),
-                userid=str(self._block.scope_ids.user_id),
-                mod_id=str(self._block.scope_ids.usage_id),
-                dispatch=dispatch,
-            ),
+            callback_type,
+            kwargs={
+                "course_id": str(course_key),
+                "userid": userid,
+                "mod_id": mod_id,
+                "dispatch": dispatch,
+            },
         )
-        xqueue_callback_url_prefix = settings.XQUEUE_INTERFACE.get('callback_url', settings.LMS_ROOT_URL)
-        return xqueue_callback_url_prefix + relative_xqueue_callback_url
+
+        xqueue_callback_url_prefix = settings.XQUEUE_INTERFACE.get("callback_url", settings.LMS_ROOT_URL)
+        return f"{xqueue_callback_url_prefix}{relative_xqueue_callback_url}"
```
Given there's only one callback now, there's no net change in this function right? This is purely refactoring and adding variable names at this point?

gabrielC1409 replied:
Yes, it is correct

ormsbee commented on xmodule/docs/decisions/0005-send-data-to-edx-submission.rst:
```diff
@@ -0,0 +1,196 @@
+
+   - **Service Selection Logic:** The file `xmodule/capa/xqueue_interfaces.py` has been updated to include a condition that checks the state of the `send_to_submission_course.enable` *waffle flag*. If the flag is active for a specific course, student responses are directed to the `edx-submission` service via the `send_to_submission` function. If inactive, the submissions continue to be processed  ...
+
+   - **Callback URL Construction:** The `construct_callback` method within the `XQueueService` class has been modified to generate the appropriate callback URL based on the state of the *waffle flag*. This ensures that evaluation results are returned correctly, whether processed by `edx-submission` or `XQueue`.
```
Please update this to note that the edx-submissions flow doesn't require a callback URL, and why.

gabrielC1409 replied:
@ormsbee, I will only remove that comment since its original function will be maintained.

ormsbee commented on xmodule/capa/xqueue_submission.py:
```diff
@@ -0,0 +1,113 @@
+
+    def _parse_json(self, data, name):
+        """
+        Helper function to safely parse data that may or may not be a JSON string.
+        This is necessary because some callers may already provide parsed Python dicts
+        (e.g., during internal calls or test cases), while other sources may send raw JSON strings.
+        This helper ensures consistent behavior regardless of input format.
+        Args:
+            data: The input to parse, either a JSON string or a Python dict.
+            name: Name of the field (used for error reporting).
+        Returns:
+            Parsed Python object or original data if already parsed.
+        Raises:
+            JSONParsingError: If `data` is a string and cannot be parsed as JSON.
```
As a general rule, we want to convert messy data at the edges, e.g. parse JSON when it first enters the system from the browser. That way, if we get invalid data, we can catch it all in one place when it enters your system, rather than it causing an error later on and deeper in the stack. Having something that sometimes takes a string and sometimes a dict can also make it harder to add type checking later. 

I'm not going to block this PR over it, but please try not to use this pattern going forward if possible.

ormsbee reviewed (changes requested):
Apologies, but could you please add the logging line here before I merge?

Thank you.
